### ISSUE TITLE
Varnish - Should only clear cache for current domain.

### ISSUE THREAD

**Issue by pynej** [magento/magento2#4871]
If you set up multiple Magneto installs and varnish and get everything working.  When you clear the Page Cache on one site it clears all magneto instances cache.  The command passed to varnish should include a host header to allow the vcl file to filter ban's on that header.

This adds the current domain to the purge request.

``` diff
--- vendor/magento/module-cache-invalidate/Model/PurgeCache.php.old 2016-06-03 14:14:45.320427000 -0400
+++ vendor/magento/module-cache-invalidate/Model/PurgeCache.php 2016-06-03 14:32:58.908971000 -0400
@@ -7,10 +7,12 @@

 use Magento\Framework\Cache\InvalidateLogger;
 use Magento\Framework\App\DeploymentConfig;
+use Magento\Framework\App\RequestInterface;

 class PurgeCache
 {
     const HEADER_X_MAGENTO_TAGS_PATTERN = 'X-Magento-Tags-Pattern';
+    const HEADER_HOST = 'host';

     /**
      * @var \Magento\PageCache\Model\Cache\Server
@@ -28,6 +30,11 @@
     private $logger;

     /**
+     * @var RequestInterface
+     */
+    private $request;
+
+    /**
      * Constructor
      *
      * @param \Magento\PageCache\Model\Cache\Server $cacheServer
@@ -37,11 +44,13 @@
     public function __construct(
         \Magento\PageCache\Model\Cache\Server $cacheServer,
         \Magento\CacheInvalidate\Model\SocketFactory $socketAdapterFactory,
-        InvalidateLogger $logger
+        InvalidateLogger $logger,
+        RequestInterface $request
     ) {
         $this->cacheServer = $cacheServer;
         $this->socketAdapterFactory = $socketAdapterFactory;
         $this->logger = $logger;
+        $this->request = $request;
     }

     /**
@@ -55,7 +64,7 @@
     {
         $socketAdapter = $this->socketAdapterFactory->create();
         $servers = $this->cacheServer->getUris();
-        $headers = [self::HEADER_X_MAGENTO_TAGS_PATTERN => $tagsPattern];
+        $headers = [self::HEADER_X_MAGENTO_TAGS_PATTERN => $tagsPattern, self::HEADER_HOST => $this->request->getHttpHost()];
         $socketAdapter->setOptions(['timeout' => 10]);
         foreach ($servers as $server) {
             try {
```

Then the vcl files can be updated to optionally filter on the domain.

``` diff
--- vendor/magento/module-page-cache/etc/varnish4.vcl.old   2016-05-16 16:18:20.000000000 -0400
+++ vendor/magento/module-page-cache/etc/varnish4.vcl   2016-06-03 15:10:56.629687000 -0400
@@ -21,6 +21,8 @@
             return (synth(400, "X-Magento-Tags-Pattern header required"));
         }
         ban("obj.http.X-Magento-Tags ~ " + req.http.X-Magento-Tags-Pattern);
+        # or only for the current domain
+        # ban("obj.http.X-Magento-Tags ~ " + req.http.X-Magento-Tags-Pattern + " && obj.http.X-Req-Host == " + req.http.host);
         return (synth(200, "Purged"));
     }

@@ -75,6 +77,8 @@
     if (bereq.url ~ "\.js$" || beresp.http.content-type ~ "text") {
         set beresp.do_gzip = true;
     }
+
+    set beresp.http.X-Req-Host = bereq.http.host;

     # cache only successfully responses and 404s
     if (beresp.status != 200 && beresp.status != 404) {
@@ -123,4 +127,5 @@
     unset resp.http.X-Varnish;
     unset resp.http.Via;
     unset resp.http.Link;
+    unset resp.http.X-Req-Host;
 }

 --- vendor/magento/module-page-cache/etc/varnish3.vcl.old  2016-05-16 16:18:20.000000000 -0400
 +++ vendor/magento/module-page-cache/etc/varnish3.vcl  2016-06-03 15:05:46.033544000 -0400
 @@ -28,6 +28,8 @@
              error 400 "X-Magento-Tags-Pattern header required";
          }
          ban("obj.http.X-Magento-Tags ~ " + req.http.X-Magento-Tags-Pattern);
 +  # or only for the current domain
 +        # ban("obj.http.X-Magento-Tags ~ " + req.http.X-Magento-Tags-Pattern + " && obj.http.X-Req-Host == " + req.http.host);
          error 200 "Purged";
      }

 @@ -85,6 +87,8 @@
          set beresp.do_gzip = true;
      }

 +    set beresp.http.X-Req-Host = bereq.http.host;
 +
      # cache only successfully responses and 404s
      if (beresp.status != 200 && beresp.status != 404) {
          set beresp.ttl = 0s;
 @@ -128,4 +132,5 @@
      unset resp.http.X-Varnish;
      unset resp.http.Via;
      unset resp.http.Link;
 +    unset resp.http.X-Req-Host;
  }
```

In short this tell magento to pass the host name onto varnish they the user can choose to use this in the varnish config.  In multi-domain sites it may cause more problems that it solves, but when running multiple sites on the same server one will not clear the cache of all others.

**Comment by choukalos** [comment]
We're tracking this improvement internally as MAGETWO-48944 .  Would be great if you can send as a PR.

**Comment by PieterCappelle** [comment]
+1

**Comment by pynej** [comment]
Create RP: https://github.com/magento/magento2/pull/4905

Note that this may not seam like a necessary feature as in some setup's varnish can look at the invoked url.  But in many cases the DNS might route externally so `http_cache_hosts` is used, or in cases with a shared caching serve or shared hosting.  In these cases the address called in varnish in not the actual site domain so it needs to be passed as a header.

The only real issue I see with this, and the reason I only added the sample code to the vcl's and didn't turn it on is in multi-site setups the current behavior may be preferred.  Ie if a single install is setup to run multiple storefronts on different domains that in that case the purge should clear all the storefronts.

**Comment by piotrekkaminski** [comment]
Thank you for your submission.

We recently made some changes to the way we process GitHub submissions to more quickly identify and respond to core code issues.

Feature Requests and Improvements should now be submitted to the new [Magento 2 Feature Requests and Improvements](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) forum (see details [here](https://community.magento.com/t5/News-Announcements/Improvements-to-GitHub-Management/m-p/44572#M96)).

We are closing this GitHub ticket and have moved your request to the new [forum](https://community.magento.com/t5/Magento-2-Feature-Requests-and/Varnish-Should-only-clear-cache-for-current-domain/idi-p/46329).

**Comment by hostep** [comment]
As far as I can see this is still missing in the latest versions of Magento. But if I'm mistaken, let me know!

In my opinion it's a bug and not a missing feature. People do deploy multiple Magento instances on one server and having a way to purge caches from Varnish for a single Magento instance is needed. Currently it looks like it purges caches for all instances of Magento which are behind the same Varnish cache.
The filtering doesn't necessarily needs to done on the domain name as being proposed by @pynej, but just on "some" identifier somehow, maybe one which can be configured in the backend of Magento which is then being used in the exported vcl file?

@ihor-sviziev: should we re-open this ticket?

@choukalos: what's the status of internal ticket MAGETWO-48944? Any progress there?

**Comment by ihor-sviziev** [comment]
Yeah, will reopen the issue

**Comment by hostep** [comment]
So, this issue got closed because nobody found it worth to confirm the issue?

**Comment by ihor-sviziev** [comment]
@hostep I believe fixing this issue will not be so trivial. I would say the priority for this issue should be P3 or even P4 because most people use varnish only with one Magento installation. 

The suggested patch will not work for multiple reasons:
- current domain != domain of the store
- for CLI runs, it will not work at all as there we don't have any 
- when we have separate domains per website, and we saved product that assigned to multiple websites - it seems as we'll have to send a separate request for each of them, that means that flushing cache will be way slower and we will need to move it to some message queue or so

If we analyze what I listed above, this issue has too low priority, fixing too small customers. It will not be a trivial fix and introduce new, not obvious issues because of the solution's raised complexity. From my perspective, we should keep this issue closed while we don't have some simple solution.

If you see there any simple solution - let us know.

**Comment by hostep** [comment]
Hi @ihor-sviziev 

We do have a bunch of smaller customers who choose to be on a server shared with some other shops, because they can't afford dedicated hosting unfortunately. I'm pretty sure we aren't the only ones dealing with such customers (Magento Cloud probably has the same problem with their integration branches where a bunch of customers share the same server resources, but I don't know that for sure)

I agree that the domain name is most likely not an adequate solution.

I don't know much about Varnish but a solution could maybe be to to have some unique identifier specified somewhere in the Magento configuration (database and/or `env.php` file) which will be send to varnish in the purge request which only removes cache entries which are associated to that unique identifier. A bit like how the default caching system in Magento works where you can assign tags to cache entries. But I don't know if that is technically possible using Varnish?

**Comment by ihor-sviziev** [comment]
@hostep seems like the use case is really rare. The author created the issue in 2016, and since that time, no one took it. It's kind of at the end of the backlog, where no one is looking because all the time will be more important. That's why I'm saying it's good to close it. Even if someone wants to fix that - he/she could still find this issue and link PR to it.
